src/preprocessing/passes/bool_to_bv.cpp

```diff
@@ -221,7 +215,157 @@ void BoolToBV::lowerNode(const TNode& n)
     default: break;
   }
 
+  // check if it's safe to lower or rebuild the node
+  // Note: might have to rebuild to keep changes to children, even if this node
+  // isn't being lowered
+
+  // it's safe to lower if all the children are bit-vectors
+  bool safe_to_lower =
+      (new_kind != k);  // don't need to lower at all if kind hasn't changed
+
+  // it's safe to rebuild if rebuilding doesn't change any of the types of the
+  // children
+  bool safe_to_rebuild = true;
+
+  for (const Node& nn : n)
+  {
+    if (safe_to_lower)
+    {
+      safe_to_lower = fromCache(nn).getType().isBitVector();
+    }
+    if (safe_to_rebuild)
+    {
+      safe_to_rebuild = (fromCache(nn).getType() == nn.getType());
+    }
+    // if it's already not safe to do either, stop checking
+    if (!safe_to_lower && !safe_to_rebuild)
+    {
+      break;
+    }
+  }
+
+  Debug("bool-to-bv") << "safe_to_lower = " << safe_to_lower
+                      << ", safe_to_rebuild = " << safe_to_rebuild << std::endl;
+
+  if (new_kind != k)
+  {
+    if (safe_to_lower)
+    {
+      rebuildNode(n, new_kind);
+      return;
+    }
+
+    else if (force && fromCache(n).getType().isBoolean())
```

Comment:
You can eliminate this `else if` my moving it's inner if into two else-if's below (adding the `new_kind != k` condition and merging the if above with its outer if.

This has the nice effect of producing an (almost) flat conditional tree with branches that correspond to the four visitation actions:
* lower to BV
* lower to BV using an ITE
* rebuild to incorporate child changes
* do nothing.

which you could then write appropriate comments for.

Reply:
Good idea, done with 03d3d6df59cfe118cf430d8483607274e60c4f11